### PR TITLE
Allow for mixing source/no-source DIFiles in one CU

### PR DESCRIPTION
**Pull request by adrian-prantl**
The DWARF proposal that the DW_LNCT_LLVM_source extension is based on
(https://dwarfstd.org/issues/180201.1.html) allows to mix source and
non-source files in the same CU by storing an empty string as a
sentinel value.

This patch implements this feature.

### CODE REVIEW

File: llvm/test/DebugInfo/Generic/mixed-source.ll [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+; CHECK-NEXT: file_names[  0]:
+; CHECK-NEXT:            name: "foo.c"
+; CHECK-NEXT:       dir_index: 0
+; CHECK-NEXT:          source: "void foo() { }\n"
+; CHECK-NEXT: file_names[  1]:
+; CHECK-NEXT:            name: "bar.h"
+; CHECK-NEXT:       dir_index: 0
+; CHECK-NOT:           source:
+
+; Ensure we reject debug info where the DIFiles of a DICompileUnit mix source
```
Comment (felipepiovezan) [inline llvm/test/DebugInfo/Generic/mixed-source.ll:13]: I believe this is a copy paste mistake from the test borrowed from the reverted commit?

**Review by felipepiovezan — APPROVED** [review]
LGTM, just have a look at the test comment

File: llvm/lib/DebugInfo/DWARF/DWARFDebugLine.cpp [hunk 2]
```diff
@@ -170,9 +170,14 @@ void DWARFDebugLine::Prologue::dump(raw_ostream &OS,
       if (ContentTypes.HasLength)
         OS << format("         length: 0x%8.8" PRIx64 "\n", FileEntry.Length);
       if (ContentTypes.HasSource) {
-        OS <<        "         source: ";
-        FileEntry.Source.dump(OS, DumpOptions);
-        OS << '\n';
+        auto Source = FileEntry.Source.getAsCString();
+        if (!Source)
+          consumeError(Source.takeError());
+        else if (*Source && (*Source)[0]) {
```
[the remark this answers is not in the document]
Comment (adrian-prantl) [reply llvm/lib/DebugInfo/DWARF/DWARFDebugLine.cpp:176]: I'll remove the safety check.

**Comment by adrian-prantl** [comment]
Landed in 87e22bdd2bd6d77d782f9d64b3e3ae5bdcd5080d.

**Pull request closed without merging**